### PR TITLE
Exclude formatting/template tokens from the per-token KL penalty (#2933)

### PR DESCRIPTION
**Pull request by behroozazarkhalili**
### Problem

With `beta != 0` (KL regularization on), formatting/template tokens (most notably the EOS/EOT token, and chat delimiters like `<|im_end|>`) can have a per-token KL 1-2 orders of magnitude higher than the semantic tokens. This spikes the KL metric and over-regularizes tokens we actually want the policy to make deterministic. See #2933 (thanks @kalomaze for the diagnosis).

### Change

Drop the tokenizer's special tokens from `per_token_kl` before it enters the loss and the metric, using `all_special_ids` (no new public config field, per @albertvillanova's note in #2933):

```python
per_token_kl = per_token_kl.masked_fill(
    torch.isin(completion_ids, self._kl_ignored_token_ids.to(completion_ids.device)), 0.0
)
```

The special-token ids are cached once in `__init__` from `self._tokenizer.all_special_ids`. These tokens still receive the policy-gradient update; only the KL pull toward the reference is removed. Applied consistently to the three duplicated K3 blocks (`grpo`, `gspo_token`, `gmpo`).

### Validation

- New test `test_kl_ignores_special_tokens`: a `beta=0.1` run trains through the masked KL path; it asserts the ignored-id cache equals the tokenizer's special ids and that params update.
- Existing `test_train_bias_correction_kl` still passes (the mask sits right after the bias-correction step).
- ruff + doc-builder clean.

### Open questions (draft, for your steer @albertvillanova)

1. **Automatic for all `beta != 0`**: this is unconditional (no config field, as you preferred). Do you want it as the default, or gated somehow?
2. **Token set**: I mask all `all_special_ids` (EOS + chat delimiters). Should it be narrower (just EOS), or is all-special the right definition of "formatting tokens"?
3. The **KL metric** now also excludes special tokens (I mask before both the loss and the logged metric). Intended, since the metric should not spike either, but flagging it.

Addresses #2933

cc @albertvillanova

---

> [!NOTE]
> **Medium Risk**
> Changes KL regularization and logged metrics for GRPO-family trainers when enabled; behavior is gated behind a default-off flag and blocked on the Liger path.
> 
> **Overview**
> Adds opt-in **`exclude_special_tokens_kl`** on `GRPOConfig` (default `False`) so EOS, pad, and chat-template delimiter tokens can be zeroed out in the per-token KL term when `beta != 0`, addressing inflated KL metrics and over-regularization on formatting tokens (#2933). Policy gradients on those positions are unchanged.
> 
> `GRPOTrainer` caches `_kl_ignored_token_ids` from `tokenizer.all_special_ids` at init and applies the mask in `_compute_loss` (after bias-correction KL when enabled), which also affects the logged `kl` metric. The same masking is applied in **GMPO** and **GSPO-token** loss paths.
> 
> **`use_liger_kernel=True`** with this flag raises `NotImplementedError` at trainer init because the fused Liger loss would skip the mask.
> 
> Tests cover training with the flag on, cache alignment with special ids, and the Liger incompatibility guard.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 7606cc2ba46bee306d84ef96f455549b0b5e1f02. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: trl/trainer/grpo_trainer.py [hunk 1]
```diff
@@ -3099,6 +3104,12 @@ def _compute_loss(self, model, inputs):
             # Importance sampling correction for the KL divergence
             if self.args.use_bias_correction_kl:
                 per_token_kl = per_token_kl * coef_1
+            # Exclude formatting/template tokens (EOS, chat delimiters) from the KL penalty: their per-token KL
+            # dominates the term and spikes the metric (issue #2933), and we want these tokens to become
+            # deterministic. They still receive the policy-gradient update below.
+            per_token_kl = per_token_kl.masked_fill(
+                torch.isin(completion_ids, self._kl_ignored_token_ids.to(completion_ids.device)), 0.0
+            )
```
[the remark this answers is not in the document]
Comment (behroozazarkhalili) [reply trl/trainer/grpo_trainer.py]: Confirmed and fixed in 98eb40d3. The finding is correct: `_kl_ignored_token_ids` was only applied inside `_compute_loss`, while `compute_liger_loss` passes `ref_per_token_logps` straight to `LigerFusedLinearGRPOLoss`, which derives the KL internally. With `use_liger_kernel=True` and `beta != 0` the mask never took effect.

Verifying it surfaced a second problem that the report did not name, and that one is arguably worse. The mask was applied unconditionally, so this PR silently changed the KL term for every GRPO run rather than only for people who want the #2933 behavior. That also ruled out reusing the `NotImplementedError` pattern from `kl_log_ratio_clip` on #6637 as-is: that flag was opt-in, so raising was free, whereas raising here would have broken every existing Liger run with `beta != 0`.

The fix therefore does both things. The exclusion now sits behind `exclude_special_tokens_kl` (default `False`, so existing runs are untouched), and combining it with `use_liger_kernel=True` raises at init instead of letting the two loss paths quietly disagree. The guard is placed with the other Liger incompatibility checks, before the liger-availability check, so it fires regardless of whether liger is installed.

Verified locally: new `test_exclude_special_tokens_kl_rejects_liger` covers the guard, `test_kl_ignores_special_tokens` covers the masked path with the flag on, and the default flag-off `beta_non_zero` tests still pass. ruff 0.13.3 (check and format) and doc-builder @2430c1e `--max_len 119` are clean.

@qgallouedec @albertvillanova flagging the API change explicitly since it turns #2933 into an opt-in flag rather than a new default. If you would rather have the more general form the issue describes (an explicit list of token ids or strings instead of a bool over `tokenizer.all_special_ids`), say so and I will switch it.

File: trl/trainer/grpo_trainer.py [hunk 2]
```diff
@@ -3099,6 +3112,13 @@ def _compute_loss(self, model, inputs):
             # Importance sampling correction for the KL divergence
             if self.args.use_bias_correction_kl:
                 per_token_kl = per_token_kl * coef_1
+            # Exclude formatting/template tokens (EOS, chat delimiters) from the KL penalty: their per-token KL
+            # dominates the term and spikes the metric (issue #2933), and we want these tokens to become
+            # deterministic. They still receive the policy-gradient update below.
+            if self.exclude_special_tokens_kl:
+                per_token_kl = per_token_kl.masked_fill(
+                    torch.isin(completion_ids, self._kl_ignored_token_ids.to(completion_ids.device)), 0.0
+                )
```
[the remark this answers is not in the document]
Comment (behroozazarkhalili) [reply trl/trainer/grpo_trainer.py:3121]: Confirmed and fixed in 7606cc2b. Verified against the files before changing anything: `gspo_token/grpo_trainer.py:60` and `gmpo/gmpo_trainer.py:117` were both still calling `masked_fill` unconditionally after the previous commit gated the main trainer.

Worth spelling out why this was quiet. Both classes subclass `GRPOTrainer` (`GRPOTrainer(_GRPOTrainer)` and `GMPOTrainer(GRPOTrainer)`), so `self._kl_ignored_token_ids` resolved fine and nothing raised. Only the semantics diverged: at the `False` default the main trainer stopped masking while the two experimental copies kept masking, so their loss and logged `kl` would silently disagree with the documented config. That is the failure mode the duplicated-code consistency rule in `AGENTS.md` exists to prevent, and my previous commit introduced it by gating one of the three K3 sites.

The gate is now applied at all three, and I checked that mechanically rather than by eye: the comment plus gate plus `masked_fill` block is byte-identical across `trl/trainer/grpo_trainer.py`, `trl/experimental/gspo_token/grpo_trainer.py` and `trl/experimental/gmpo/gmpo_trainer.py`.

Verified: `pytest tests/experimental -k "gmpo or gspo"` gives 14 passed, including `test_train_with_kl` for both trainers. ruff 0.13.3 (check and format) and doc-builder @2430c1e `--max_len 119` are clean on both files.

**Comment by qgallouedec** [comment]
closing see https://github.com/huggingface/trl/issues/2933#issuecomment-5404693067
